### ISSUE TITLE
Dynamically sized arrays with maximum length

### ISSUE THREAD

**Issue by Justin-Tan** [google/jax#3345]
Jax doesn't seem to like dynamically sized arrays when `jit`ing, which is reasonable, because I suppose static shapes make it easier for the compiler? e.g. the following yields an error:

```
import jax.numpy as jnp
from jax import random
from jax import jit

rng = random.PRNGKey(0)
K = random.randint(rng, (1,), 1, 100) 

def dynamic_array(K):
   return jnp.arange(K)

# Raises an error
jit(dynamic_array)(K)
```
Similar considerations exists if using `lax.iota` or `linspace`.  Is there a way to specify some maximum array size and getting away with creating an array of dynamic length below this maximum size? I thought about allocating an array of maximum length and using `lax.dynamic_slice`, but this throws a similar error, and is pretty wasteful.

Right now I get around this by specifying the size `K` as a static argnum to `jit`, but then this recompiles for each different value of `K`, if I understand correctly, and seems pretty wasteful if there can be many different sizes `K`.

For some context, I'm doing this in the context of Russian Roulette estimators - (https://projecteuclid.org/euclid.ss/1449670853), where one can obtain an unbiased estimate of some infinite series by randomly truncating the series and weighting terms appropriately.

**Comment by pedrofale** [comment]
Any progress on this? This would also be useful for Bayesian nonparametric modelling.

**Comment by machineko** [comment]
@mattjj Would love to hear some update if u have any 🍨

**Comment by sooheon** [comment]
AFAICT this also applies for graph NNs without padding a la https://github.com/google/flax/blob/84674691f91c210dcdbe0c144f7b37c3800fd22a/examples/graph/train.py#L60

**Comment by mattjj** [comment]
Sorry for not providing any useful updates. Actually I didn't notice until now that this issue was getting pinged! My GitHub notifications are not organized, to say the least.

We've worked on a couple prototypes in this direction, but no solutions so far. Momentum is picking back up on this front so I hope to provide a substantial update in the next several months. (I'd love to detail and discuss all our work-in-progress ideas here, but we're pretty bandwidth-maxed right now!)

So I don't have a useful update right now, but we haven't forgotten!

**Comment by aniquetahir** [comment]
Any updates? I was trying to re-implement IRM with CMNIST in Jax but kind of impossible to reproduce the original version because of this issue.

**Comment by aniquetahir** [comment]
> May I know if there is any update? This is a blocking issue of using JAX in dynamic environment.

One solution I ended up using is to make the shape as an additional parameter with static argname/argnum. If the number of shapes is constant, it will just recompile the function for each unique shape.
